src/data/polynomial/cyclotomic_polynomials.lean

```diff
@@ -0,0 +1,367 @@
+varies over the `n`-th roots of unity. -/
+lemma X_pow_sub_one_eq_prod {ζ : K} {n : ℕ} (hpos : 0 < n) (h : is_primitive_root ζ n) :
+  X ^ n - C (1 : K) = ∏ ζ in nth_roots_finset n K, (X - C ζ) :=
+begin
+  rw [nth_roots_finset, ← multiset.to_finset_eq (is_primitive_root.nth_roots_nodup h)],
+  simp only [finset.prod_mk, ring_hom.map_one],
+  rw [nth_roots],
+  have hmonic : (X ^ n - C (1 : K)).monic := monic_X_pow_sub_C (1 : K) (ne_of_lt hpos).symm,
+  symmetry,
+  apply prod_multiset_X_sub_C_of_monic_of_roots_card_eq hmonic,
+  rw [@nat_degree_X_pow_sub_C K _ _ n hpos 1, ← nth_roots],
+  exact is_primitive_root.card_nth_roots h
+end
+
+/-- `cyclotomic' n K` splits. -/
+lemma cycl_splits (n : ℕ) : splits (ring_hom.id K) (cyclotomic' n K) :=
+begin
+  apply splits_prod (ring_hom.id K),
+  intros z hz,
+  simp only [splits_X_sub_C (ring_hom.id K)]
+end
+
+/-- If there is a primitive `n`-th root of unity in `K`, then `X ^ n - 1`splits. -/
+lemma X_pow_sub_one_splits {ζ : K} {n : ℕ} (h : is_primitive_root ζ n) :
+  splits (ring_hom.id K) (X ^ n - C (1 : K)) :=
+begin
+  by_cases hzero : n = 0,
+  { simp only [hzero, ring_hom.map_one, splits_zero, pow_zero, sub_self] },
+  apply (splits_iff_card_roots (X_pow_sub_C_ne_zero (nat.pos_of_ne_zero hzero) (1 : K))).2,
+  rw [← nth_roots, is_primitive_root.card_nth_roots h, nat_degree_X_pow_sub_C],
+  exact nat.pos_of_ne_zero hzero
+end
+
+/-- If there is a primitive `n`-th root of unity in `K`, then
+`X ^ n - 1 = ∏ (cyclotomic' i K)`, where `i` divides `n`. -/
+lemma X_pow_sub_one_eq_prod_cycl' {ζ : K} {n : ℕ} (hpos : 0 < n) (h : is_primitive_root ζ n) :
+  X ^ n - C (1 : K) = ∏ i in nat.divisors n, cyclotomic' i K :=
+begin
+  rw [X_pow_sub_one_eq_prod hpos h],
+--  rw ← pnat.to_pnat'_coe hpos at h ⊢,
+--  set m := nat.to_pnat' n,
+  have rwcyc : ∀ i ∈ nat.divisors n, cyclotomic' i K = ∏ μ in primitive_roots i K, (X - C μ),
+  { intros i hi,
+    simp only [cyclotomic'] },
+  conv_rhs { apply_congr,
+             skip,
+             simp [rwcyc, H] },
+  rw ← finset.prod_bind,
+  { simp only [is_primitive_root.nth_roots_one_eq_bind_primitive_roots hpos h] },
+  intros x hx y hy hdiff,
+  simp only [nat.mem_divisors, and_true, ne.def, pnat.ne_zero, not_false_iff] at hx hy,
+  refine is_primitive_root.disjoint _ _ hdiff,
+  exact @nat.pos_of_mem_divisors n x (nat.mem_divisors.2 hx),
+  exact @nat.pos_of_mem_divisors n y (nat.mem_divisors.2 hy)
+end
+
+/-- If there is a primitive `n`-th root of unity in `K`, then `cyclotomic' n K` comes from a
+polynomial with integer coefficients. -/
+lemma int_coeff_of_cycl {ζ : K} {n : ℕ} (h : is_primitive_root ζ n) :
+  (∃ (P : polynomial ℤ), map (int.cast_ring_hom K) P = cyclotomic' n K ∧
+  P.degree = (cyclotomic' n K).degree ∧ P.monic) :=
+begin
+  refine lifts_and_degree_eq_and_monic _ (cyclotomic'.monic n K),
+  revert h ζ,
+  apply nat.strong_induction_on n,
+  intros k hk z hzeta,
+  by_cases hzero : k = 0,
```

Comment:
Again you could use `cases nat.eq_zero_or_pos k with hzero hpos` and then you don't need `have hpos := (nat.pos_of_ne_zero hzero),` four lines later, and you don't have `hzero : not (k = 0)` clogging up your local context.